File: ios-native/Dimo/Email/Gmail/GmailCredentialVault.swift
import CryptoKit
import Foundation
import Security

struct ConnectedGmailAccount: Codable, Hashable, Sendable, Identifiable {
  var subject: String
  var emailAddress: String
  var connectedAt: Date

  var id: String { subject }
}

struct GmailStoredCredential: Codable, Hashable, Sendable {
  var subject: String
  var emailAddress: String
  var refreshToken: String
  var connectedAt: Date

  var account: ConnectedGmailAccount {
    ConnectedGmailAccount(
      subject: subject,
      emailAddress: emailAddress,
      connectedAt: connectedAt
    )
  }
}

private struct GmailCredentialBundle: Codable, Sendable {
  static let currentVersion = 1

  var version: Int
  var credentials: [GmailStoredCredential]
}

/// Stores Gmail refresh tokens in a device-only, Dimo-user-scoped Keychain item.
/// Access tokens are intentionally absent from this type and remain in memory.
actor GmailCredentialVault {
  private static let service = "app.dimo.ios.gmail"
  private let encoder: JSONEncoder
  private let decoder: JSONDecoder

  init() {
    encoder = JSONEncoder()
    encoder.dateEncodingStrategy = .millisecondsSince1970
    decoder = JSONDecoder()
    decoder.dateDecodingStrategy = .millisecondsSince1970
  }

  func accounts(dimoUserId: String) throws -> [ConnectedGmailAccount] {
    try loadBundle(dimoUserId: dimoUserId).credentials
      .map(\.account)
      .sorted { $0.emailAddress.localizedCaseInsensitiveCompare($1.emailAddress) == .orderedAscending }
  }

  func credential(subject: String, dimoUserId: String) throws -> GmailStoredCredential? {
    try loadBundle(dimoUserId: dimoUserId).credentials.first { $0.subject == subject }
  }

  func upsert(_ credential: GmailStoredCredential, dimoUserId: String) throws {
    var bundle = try loadBundle(dimoUserId: dimoUserId)
    bundle.credentials.removeAll { $0.subject == credential.subject }
    bundle.credentials.append(credential)
    try save(bundle, dimoUserId: dimoUserId)
  }

  func remove(subject: String, dimoUserId: String) throws {
    var bundle = try loadBundle(dimoUserId: dimoUserId)
    bundle.credentials.removeAll { $0.subject == subject }
    if bundle.credentials.isEmpty {
      try deleteBundle(dimoUserId: dimoUserId)
    } else {
      try save(bundle, dimoUserId: dimoUserId)
    }
  }

  func removeAll(dimoUserId: String) throws {
    try deleteBundle(dimoUserId: dimoUserId)
  }

  private func loadBundle(dimoUserId: String) throws -> GmailCredentialBundle {
    let query: [String: Any] = [
      kSecClass as String: kSecClassGenericPassword,
      kSecAttrService as String: Self.service,
      kSecAttrAccount as String: accountKey(for: dimoUserId),
      kSecReturnData as String: true,
      kSecMatchLimit as String: kSecMatchLimitOne,
    ]
    var item: CFTypeRef?
    let status = SecItemCopyMatching(query as CFDictionary, &item)
    if status == errSecItemNotFound {
      return GmailCredentialBundle(version: GmailCredentialBundle.currentVersion, credentials: [])
    }
    guard status == errSecSuccess, let data = item as? Data else {
      throw GmailCredentialVaultError.keychain(status)
    }
    let bundle: GmailCredentialBundle
    do {
      bundle = try decoder.decode(GmailCredentialBundle.self, from: data)
    } catch {
      throw GmailCredentialVaultError.corruptBundle
    }
    guard bundle.version == GmailCredentialBundle.currentVersion else {
      throw GmailCredentialVaultError.unsupportedBundleVersion(bundle.version)
    }
    return bundle
  }

  private func save(_ bundle: GmailCredentialBundle, dimoUserId: String) throws {
    let data = try encoder.encode(bundle)
    let baseQuery: [String: Any] = [
      kSecClass as String: kSecClassGenericPassword,
      kSecAttrService as String: Self.service,
      kSecAttrAccount as String: accountKey(for: dimoUserId),
    ]
    let updateStatus = SecItemUpdate(
      baseQuery as CFDictionary,
      [kSecValueData as String: data] as CFDictionary
    )
    if updateStatus == errSecSuccess { return }
    guard updateStatus == errSecItemNotFound else {
      throw GmailCredentialVaultError.keychain(updateStatus)
    }
    var insert = baseQuery
    insert[kSecValueData as String] = data
    insert[kSecAttrAccessible as String] = kSecAttrAccessibleAfterFirstUnlockThisDeviceOnly
    let addStatus = SecItemAdd(insert as CFDictionary, nil)
    guard addStatus == errSecSuccess else {
      throw GmailCredentialVaultError.keychain(addStatus)
    }
  }

  private func deleteBundle(dimoUserId: String) throws {
    let query: [String: Any] = [
      kSecClass as String: kSecClassGenericPassword,
      kSecAttrService as String: Self.service,
      kSecAttrAccount as String: accountKey(for: dimoUserId),
    ]
    let status = SecItemDelete(query as CFDictionary)
    guard status == errSecSuccess || status == errSecItemNotFound else {
      throw GmailCredentialVaultError.keychain(status)
    }
  }

  private func accountKey(for dimoUserId: String) -> String {
    let digest = SHA256.hash(data: Data(dimoUserId.utf8))
    return "gmail.credentials." + digest.map { String(format: "%02x", $0) }.joined()
  }
}

enum GmailCredentialVaultError: LocalizedError {
  case keychain(OSStatus)
  case corruptBundle
  case unsupportedBundleVersion(Int)

  var errorDescription: String? {
    switch self {
    case .keychain(let status): return "Gmail credentials could not be accessed (\(status))."
    case .corruptBundle: return "The saved Gmail credentials are corrupt."
    case .unsupportedBundleVersion(let version):
      return "The saved Gmail credential format (\(version)) is unsupported."
    }
  }
}

File: android-native/app/src/main/java/app/dimo/android/email/gmail/GmailCredentialVault.kt
package app.dimo.android.email.gmail

import android.content.Context
import android.content.SharedPreferences
import androidx.security.crypto.EncryptedSharedPreferences
import androidx.security.crypto.MasterKey
import java.security.MessageDigest
import java.util.concurrent.ConcurrentHashMap
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json

@Serializable
data class ConnectedGmailAccount(
  val subject: String,
  val emailAddress: String,
  val connectedAt: Long,
) {
  val id: String get() = subject
}

@Serializable
data class GmailStoredCredential(
  val subject: String,
  val emailAddress: String,
  val refreshToken: String,
  val connectedAt: Long,
) {
  val account: ConnectedGmailAccount
    get() = ConnectedGmailAccount(subject, emailAddress, connectedAt)
}

@Serializable
private data class GmailCredentialBundle(
  val version: Int,
  val credentials: List<GmailStoredCredential>,
) {
  companion object {
    const val CURRENT_VERSION = 1
  }
}

/**
 * Stores Gmail refresh tokens in a device-only, Dimo-user-scoped encrypted file —
 * the Keystore counterpart of the Keychain vault in
 * `ios-native/Dimo/Email/Gmail/GmailCredentialVault.swift`.
 *
 * Access tokens are intentionally absent from this type and remain in memory.
 * Every accessor is guarded by a mutex so a concurrent connect and refresh cannot
 * lose a credential through read-modify-write.
 */
class GmailCredentialVault(context: Context) {
  private val appContext = context.applicationContext
  private val mutex = Mutex()
  private val json = Json { ignoreUnknownKeys = true }
  private val files = ConcurrentHashMap<String, SharedPreferences>()

  suspend fun accounts(dimoUserId: String): List<ConnectedGmailAccount> = mutex.withLock {
    loadBundle(dimoUserId).credentials
      .map { it.account }
      .sortedBy { it.emailAddress.lowercase() }
  }

  suspend fun credential(subject: String, dimoUserId: String): GmailStoredCredential? =
    mutex.withLock {
      loadBundle(dimoUserId).credentials.firstOrNull { it.subject == subject }
    }

  suspend fun upsert(credential: GmailStoredCredential, dimoUserId: String) = mutex.withLock {
    val bundle = loadBundle(dimoUserId)
    save(
      bundle.copy(
        credentials = bundle.credentials.filter { it.subject != credential.subject } +
          credential,
      ),
      dimoUserId,
    )
  }

  suspend fun remove(subject: String, dimoUserId: String) = mutex.withLock {
    val bundle = loadBundle(dimoUserId)
    val remaining = bundle.credentials.filter { it.subject != subject }
    if (remaining.isEmpty()) {
      deleteBundle(dimoUserId)
    } else {
      save(bundle.copy(credentials = remaining), dimoUserId)
    }
  }

  suspend fun removeAll(dimoUserId: String) = mutex.withLock { deleteBundle(dimoUserId) }

  // MARK: - Private

  private fun loadBundle(dimoUserId: String): GmailCredentialBundle {
    val raw = prefs(dimoUserId).getString(KEY_BUNDLE, null)
      ?: return GmailCredentialBundle(GmailCredentialBundle.CURRENT_VERSION, emptyList())
    val bundle = runCatching { json.decodeFromString<GmailCredentialBundle>(raw) }.getOrNull()
      ?: throw GmailCredentialVaultException("The saved Gmail credentials are corrupt.")
    if (bundle.version != GmailCredentialBundle.CURRENT_VERSION) {
      throw GmailCredentialVaultException(
        "The saved Gmail credential format (${bundle.version}) is unsupported.",
      )
    }
    return bundle
  }

  private fun save(bundle: GmailCredentialBundle, dimoUserId: String) {
    prefs(dimoUserId).edit().putString(KEY_BUNDLE, json.encodeToString(bundle)).apply()
  }

  private fun deleteBundle(dimoUserId: String) {
    prefs(dimoUserId).edit().clear().apply()
  }

  /**
   * One encrypted file per Dimo user. The name is a digest so a WorkOS user id
   * never lands in a filename on disk.
   */
  private fun prefs(dimoUserId: String): SharedPreferences = files.getOrPut(dimoUserId) {
    val masterKey = MasterKey.Builder(appContext)
      .setKeyScheme(MasterKey.KeyScheme.AES256_GCM)
      .build()
    EncryptedSharedPreferences.create(
      appContext,
      fileName(dimoUserId),
      masterKey,
      EncryptedSharedPreferences.PrefKeyEncryptionScheme.AES256_SIV,
      EncryptedSharedPreferences.PrefValueEncryptionScheme.AES256_GCM,
    )
  }

  private fun fileName(dimoUserId: String): String {
    val digest = MessageDigest.getInstance("SHA-256")
      .digest(dimoUserId.toByteArray(Charsets.UTF_8))
    return FILE_PREFIX + digest.joinToString("") { "%02x".format(it) }
  }

  private companion object {
    const val FILE_PREFIX = "dimo.gmail."
    const val KEY_BUNDLE = "gmail.credentials"
  }
}

class GmailCredentialVaultException(message: String) : Exception(message)
